Pick/Drillpick Imagery Layer

srothst1 opened the pull request:
Fixes #8692

Based on the discussion from [this post](https://github.com/CesiumGS/cesium/issues/8692), there is clearly demand for functionality allowing users to select an imagery layer (or imagery layers) from our viewer. Currently, in demos [like this one](http://localhost:8080/Apps/Sandcastle/index.html#c=1Vlbb+O4Ff4rhB92bYxDObsY7G5ubSYZpAaSZhFndlCMg4CWaJsTilRJyo678H/v4UUSZSvZyfah7cuMRR6ey3cuPDxZEYVWjK6pQqdI0DW6oJqVOf7NrfWnvdR9X0hhCBNUTXtD9PtUIDQjml6TDVW/svSJqiM0J1zT4VRsB8dTsQK2LCcLqjaOSAN3Lwa3loHUE9vNG5lRDoSOP3f7R+jLw7AlrlkqC7dwhETJuVvJ5FrsrmnKaWpotrvO9MTtkBmnYW9eitQwKVDfCR94RRBS1JRKILNkGjd6YCYy+nw7r4jPTtHo2B7YOv6KME33eA6RO1WzboGB3ZnA79gT1LjgYC7g4wj29mvjA9R+1cP45YaYJc7Jc38UNEAH6HDwEJiURUaMN+uaadOvhK+BVK6xpuae5VSWpt+YU5vQrWO3Xhb9wHs7RIejQQQYl+sOJ7wOmDvzh4Dt6xlQqTYQ8vgw0d+lwZyKhVlauCrk3gFy1cmHV90QO+p/AuOUiLvOuBy3QA4B3+ygs1ZsA5vrTm/9IRvIEfTdd/HSya5La7xrgdtQUJrcaxnfLLtqEgrYk5DpE8CJjSLpU+PXgSOq9QbcA6C6gjtJ0IWi4CxEOIesp05wKEjwTYwrBYg+k7zgFDwIZLosCqkMDgzul1S3zxFFxfcGMAGmG6QLmjLCMUJj871GhdSaQSFCRqIlWVGUl9wwy1zOrQZ5YEuFLVcZAhWkSOkQfS21QZw9UaenhH9UEDhEs9IgZqAoUm0l5+TJ8k2XYLOwytGUlJoGztZUL6tROZUrYGcZU2GYomjB5QxQEZm1BsmC/LOkzmKSZR8qL9wWFli4OT4wsYDEKjQ6pwqMtVdHCT6fwz2SDY6tVMs7LRVAYyKwXuLogirmeydJBlz9RvB76lz3WSqejX1g9esk0mbDIfYD5ViKmGxiN/Hd7fllYLh1We7T5xV1zlV6NZ4gxwpNjKLUOO1qxaJb1dIyDbsTqgDbIPlXJVcsg0JWK1oqftQUp2lvaUyhj5JEwzEGNzImKl0wLQUHLHEq88RrkSiqTU2VOJ0evU6PIDWpJdfKfaORtwUVns9LprUodg37NiETQ3Iq0J+S0IIuRszxPDCMW9Sw1hzPiTZ8gwU1yRpCRaWSS5XUEhGaA/HHZwNZAtpZZl+LRbQNEZYx0/IPKIScCDTboGDGJWi9ED7kFbq4QB/+gX7EI4wuiSGWrmVPm25yjt/qob8TKLaEo49EQfkcj6Eiy5TwQReS96BqCAWIk9ehDEdmJeMZFNCS00+KQ3afayidOrmnzyAXYi3Id+LH42lv8Db1P02uJmiyJBlUtzvKGZ0jqNkEfb65n3Ta8JnOwARrybeY0ZVMtuDkpMCCWE0It78XcpX43NpJJaufV89rdyv4pkmnxKoZhQj3DaU3a/dYRBcK0rQHVZFAyY+DUKqcGLvn7sfka0HjIDTOhUax5wk148uIx+iHn/M8ooSuj+Vlfk1XFNx5+MtuHFstMZpgdEUhERYMggZNSrBq0xGC7fvRlm/wKfP4hWsjOPq8Xu90t2DQk9tUMZA0V7cfJ2gs5gqulexlZ/9ZR+cUKiU1mCyUFJhB+oNUTLMyAUcfzJhI1rlOFnBPJqkUpX5kCsPOX3Ydqi1WluyxItuvCkBRGYIym+ipLBXYuEFjuSboo1gxUALqgwHAbrxiEZeCKCgexsn6vbEFehihC9tCOAlGlTQ6tB8shYhjZdudim/x0Gfwt+0t7khG1H/LQ4I+K5IlYvSKezzNAdB0+sbp///tmHbxRg7r/7TIT3sYJ05DSAF3/BGqoa0xKvHsHq/lQj5e2KsyKgvhxwj/8AYDJtC/8VcU9/tW/W9XOlbSdq0QELiNtbJvfcu4vtLuqhU8VzK/pAu4inW/8d3B4eF7PIqc+ePPrc+Dw9FP8e4v+Kf31edgF6JDPHoDRFeKddZBu77XWzUSwi83iXljRKELKVXGhM32l6IpInmjEtvWm6ujDRCQXMPqJVhxDe8x+/RrXtIMGgOzKSg8VnbI0enpKVhTPzGgB6lSlYe3cfupuaCmP9qfG7QmRq1X/BZRMGqXaYTSOOLf37XGs/B3qDuLrdWWAfx33B5y4aLUy2awsYdft0M7URzC265YkiHSS7nehXQPE+C969tOO7wBjjPwCOaHDuQ3wkvaD1JH+H18wirxwgG7NUS2hMYHdjHqft6HWdGXac+JtU/Nac8y9L/s6WnvYR/KvalMg4+AnqkaNnSNKJw2e+MiXXCotnbM9sIkydsG9wL0t3ZI6iY2lSQ79YAlN0pEBwesDuA9Zi4+9sOZDeowc4a2xhuw0zWFmookgTd15tq5Zg5rpFu4vL2BqKf2YtRhQoA+jf0YJLVTA2sKs4MPa46Rks+IDatMpqU9ZNX66M9/2IwzeDkEGkjP4/1pDSkKvrHawCWgm6nNsOI86JrxWIutnNuZ7dftiCQ+CYEQZzTIdfS6nOlUsRmNBm51AtbAQ7P7NyIyKJDpEi4LaIUCLpmSxYGdu0WDizBslgoHrwESBFivPFM/8KrGxPtR9kLEBOpWyIxsmJw0p4/Ru3csGhcy93yq2O3NuvfGnF/Yw2AQNzYdarPjZnsGD4Cn+nsbLrtOq7vG0bvcHyJEQoWISFzRiCiqqhOTuLXjzqE6zaEj6EfEQ385DTrJoQA2YTBs5eYeJnaGHpjUR6oKF+nc7NWaR9p2D4a3PtB7w96JeySeVVj/leV20GhboD70P4bmBbc3cjIr0ydooVOtK51OkvjoScZWiGWnHX/SQSknWsPOvOR8wv4FtfLsJAH6vaNcEpuZt77FsmTLw7Nrv4gxPkngs/tknfWt/X8D), if a user selects an imagery layer and not a feature they are presented with the message "No features found." The final goal of this pull request is to give users easy access to imagery layer data.

I have spent the last few days looking through `ImageryLayer.js`, `ImageryLayerCollection.js` and `Viewer.js`. I believe that this functionality should be added to a new method `pickImageryLayer` in `ImageryLayerCollection.js` and eventually integrated into either `viewer.pickEntity` or `viewer.pickImageryLayerFeature`.

[Here is a sandcastle demo](http://localhost:8080/Apps/Sandcastle/index.html#c=tVVNb9s4EP0rs75ERgxye3XcoEGQdoOmaVEH28O6CGhqbBGlSIGkHKiF/3uHoiTHSrLYHPZgi5qP94YzT+ROONgpfEAHb8HgA1yiV3XJ/m5t2Woi2/dLa4JQBt1qMoNfKwOgSrFF13xxdqdydPPHyRdOflD+k6iW6Hboro9DszYfoHZ6nlYAq0kRQuXnnHtKUUTKhJNb5a3RRMukLXlEvV5yhz4MUfxWhA9o779Zp/N7YuQDK6dSI/x+So/99GxlVobzC+0tBNeAq41RWwiF8uCFyaXwQSM8qFAkY+rKDB4KJQsohAdj+12DFg06D5QXrRsUoaa6IDLsXt3RVFvKYV46RLOshMSrHZrwF3HoaMdwbao6XMigrMk2tWkXkJW29jhNQxkwkHpWKfkjeVllvYrRkYmiqEInGiqvi5eiRCfYFsMXyvkqmhfTttqu8ZCYiFrjECKFC7RDYSis9aRCiHB2lJZA1QayP7oO5bihhuTZgDDttgXgkBps2ox9/Is/zm+vvs0honcKu2mnEn3SGm81Mm231PHODck/h9TyVG4/0WFP6jEWe4L+mo10UIdtPC7r5PYgp8iCOTtJUHtA7bHP4Rzuoh6HidNaCq0xB+EbIwtnDU1LN6RUNKRdBK18iAUl2IM8KVPshNJiTTU8V1Fdrkm2djPS+RxO4LS3MY1mG4pUKsDGOsg0Eh+18M8zeixGkWQ7PZ3+6j/1I8a2p8fwpMJMTXv4/Wjk43m87/b2dOi953jcfS/oMCqVx/869h7sNdMfUb2sgmE+/yaDDjzOeIw8G7QxeAau/1s/LypoM7Q/Drd/G4nnefmMgkf6GTF2kzmm+Ud9Z4YOtQPPvpPTtD9D6E6Y9U1djs7cu6ZCdnP1/u7+8ub68uP0bDKbLHxoNJ73eO9UWVkX4iWWMcYDlpUWdGrxdU09C0x635Mv+OPURa52oPK3z9wDILXwnjybWuul+omryfmCU/yTVG1Frsz2M11z9IXGsOLN+U0yMsYWnF6fzwzW6rVwI+Tf) that showcases how this functionality currently works. When selecting a position on the globe, we see that `pickImageryLayers` returns an `ImageryLayer` and `pickImageryLayerFeatures` returns nothing since there are no features. I think this is a step in the right direction.

While I am looking for overall feedback and clarification, I do have some more specific questions:
- Should `ImageryLayerCollection.prototype.pickImageryLayers` return a `ImageryLayerCollection` or a promise to `ImageryLayerFeatureInfo`?
- How should `ImageryLayerCollection.prototype.pickImageryLayers` be integrated into `ImageryLayer.js` and `Scene.js`?

When I get a better understanding of what needs to be done, I will update `ImageryLayerCollectionSpec.js` and `CHANGES.md`. I do not expect this to be part of Thursday's release.

@lilleyse 
@ebogo1

cesium-concierge commented:
Thanks for the pull request @srothst1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/drillpick-imagery-layer/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -325,6 +325,105 @@ ImageryLayerCollection.prototype.lowerToBottom = function (layer) {
+ *
+ * @param {Ray} ray The ray to test for intersection.
+ * @param {Scene} scene The scene.
+ * @return {ImageryLayerCollection|undefined} An `ImageryLayerCollection` that includes all of
```
I like the thought of returning an `ImageryLayerCollection` but I think an array of `ImageryLayer`'s is fine. This matches `drillPick` and `pickImageryLayerFeatures` more closely.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -325,6 +325,105 @@ ImageryLayerCollection.prototype.lowerToBottom = function (layer) {
+    if (!defined(imagery)) {
+      continue;
+    }
+    var provider = imagery.imageryLayer.imageryProvider;
+    if (!defined(provider.pickFeatures)) {
+      continue;
+    }
+
+    if (!Rectangle.contains(imagery.rectangle, pickedLocation)) {
+      continue;
+    }
+
+    // If this imagery came from a parent, it may not be applicable to its entire rectangle.
+    // Check the textureCoordinateRectangle.
+    var applicableRectangle = applicableRectangleScratch;
+
+    var epsilon = 1 / 1024; // 1/4 of a pixel in a typical 256x256 tile.
+    applicableRectangle.west = CesiumMath.lerp(
+      pickedTile.rectangle.west,
+      pickedTile.rectangle.east,
+      terrainImagery.textureCoordinateRectangle.x - epsilon
+    );
+    applicableRectangle.east = CesiumMath.lerp(
+      pickedTile.rectangle.west,
+      pickedTile.rectangle.east,
+      terrainImagery.textureCoordinateRectangle.z + epsilon
+    );
+    applicableRectangle.south = CesiumMath.lerp(
+      pickedTile.rectangle.south,
+      pickedTile.rectangle.north,
+      terrainImagery.textureCoordinateRectangle.y - epsilon
+    );
+    applicableRectangle.north = CesiumMath.lerp(
+      pickedTile.rectangle.south,
+      pickedTile.rectangle.north,
+      terrainImagery.textureCoordinateRectangle.w + epsilon
+    );
+    if (!Rectangle.contains(applicableRectangle, pickedLocation)) {
+      continue;
+    }
```
Since there is a lot of duplication with `pickImageryLayerFeatures` see if you can factor this into a common function that `pickImageryLayers` and `pickImageryLayerFeatures` can both use.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -325,6 +325,105 @@ ImageryLayerCollection.prototype.lowerToBottom = function (layer) {
 var applicableRectangleScratch = new Rectangle();
 
+/**
+ * Determines the imagery layers that are intersected by a pick ray. To compute a pick ray from a
+ * location on the screen, use {@link Camera.getPickRay}.
```
I posted some caveats about the rectangle-based approach in https://github.com/CesiumGS/cesium/issues/8692. This function is still helpful in a lot of cases but the limitations should be documented.

lilleyse reviewed (commented):
Not too many comments on the code itself which looks pretty good. The major limitation of this approach is that it isn't pixel perfect. Still I think it's would be valuable until someone implements the pixel perfect approach. I wrote some starter notes on that here: https://github.com/CesiumGS/cesium/issues/8692#issuecomment-873488501.

Probably not worth tackling now unless you are up for it.

srothst1 commented:
@lilleyse Thank you for the feedback. I am happy to hear that the code looked mostly good. I implemented some of your suggestions:
1. `pickImageryLayers` returns an array of `ImageryLayer` objects instead of an `ImageryLayerCollection`. Following the precedent set by `drillPick` and `pickImageryLayerFeatures`.
2. Created a helper function (`pickImageryLayersHelper`) for `pickImageryLayers` and `pickImageryLayerFeatures`, removing a lot of duplicated code.
3. Updated the documentation to include information about the caveats of the rectangle-based approach currently implemented.

Before I consider implementing a pixel-perfect version of `pickImageryLayers`, I want to ensure that the code I currently have is correct.

A few questions:
1. How do I generate the doc to make sure everything looks okay?
2. Should I generate a spec for `pickImageryLayersHelper` or `pickImageryLayers`?

srothst1 commented:
@lilleyse Thank you for the feedback. I am happy to hear that the code looked mostly good. I implemented some of your suggestions:
1. `pickImageryLayers` returns an array of `ImageryLayer` objects instead of an `ImageryLayerCollection`. Following the precedent set by `drillPick` and `pickImageryLayerFeatures`.
2. Created a helper function (`pickImageryLayersHelper`) for `pickImageryLayers` and `pickImageryLayerFeatures`, removing a lot of duplicated code.
3. Updated the documentation to include information about the caveats of the rectangle-based approach currently implemented.

Before I consider implementing a pixel-perfect version of `pickImageryLayers`, I want to ensure that the code I currently have is correct.

A few questions:

1. On my local machine, [all specs](http://localhost:8080/Specs/SpecRunner.html?) pass. Why are some checks not successful on GitHub? How should I debug when this occurs?
2. How do I generate the documentation to make sure everything is generated properly?
3. Should I create a spec for `pickImageryLayersHelper` or `pickImageryLayers`?

lilleyse commented:
1. It could be a fluke. I restarted the build and CI passed. If you see similar issues open an issue.
2. Run `npm run generateDocumentation`. If you're curious about the various build commands see the [Build Guide](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/BuildGuide)
3. Add tests for `pickImageryLayers`. It's fine not to test `pickImageryLayersHelper`. On the topic of tooling, you can check how well the unit tests cover the the code by running `npm run coverage` and seeing the coverage results. If you're only concerned with a single suite of tests you can change `describe` to `fdescribe` and it will only run those tests.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -389,6 +371,136 @@ ImageryLayerCollection.prototype.pickImageryLayerFeatures = function (
+ *
+ * @param {Ray} ray The ray to test for intersection.
+ * @param {Scene} scene The scene.
+ * @return {ImageryLayer[]|undefined} An `ImageryLayerCollection` that includes all of
+ *                                 the layers that are intersected by a given pick ray. Undefined if
+ *                                 no layers are selected.
```
```suggestion
 * @return {ImageryLayer[]|undefined} An array that includes all of
 *                                 the layers that are intersected by a given pick ray. Undefined if
 *                                 no layers are selected.
```

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -389,6 +371,136 @@ ImageryLayerCollection.prototype.pickImageryLayerFeatures = function (
+      continue;
+    }
+    var provider = imagery.imageryLayer.imageryProvider;
+    if (!defined(provider.pickFeatures)) {
+      continue;
+    }
```
This check is is not needed for `pickImageryLayers` since `pickImageryLayers` isn't concerned about the features.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -326,38 +326,20 @@ ImageryLayerCollection.prototype.lowerToBottom = function (layer) {
- * }
  */
-ImageryLayerCollection.prototype.pickImageryLayerFeatures = function (
+ImageryLayerCollection.prototype.pickImageryLayersHelper = function (
```
In most cases helper functions don't need to be added to the class, they can just exist as local functions in the file.

lilleyse commented:
The new approach is better but there's still a lot of code duplication. One idea is to pass a callback function to the helper so that once overlapping imagery is found it can let the callback do the rest of the work. This lets `pickImageryLayers` and `pickImageryLayerFeatures` define their own custom behavior with little code duplication.

It was a bit hard to annotate the idea on github so I mocked something up: [callback.txt](https://github.com/CesiumGS/cesium/files/6793902/callback.txt). I haven't actually tested it so double check that there aren't any mistakes. What do you think?

lilleyse commented:
> On my local machine, all specs pass. Why are some checks not successful on GitHub? How should I debug when this occurs?

I submitted an issue for this https://github.com/CesiumGS/cesium/issues/9672

srothst1 commented:
@lilleyse Thanks for the feedback! I added a series of commits. Here is a quick summary:
1. Added tests for `pickImageryLayers`. I modeled these tests after the tests for `pickImageryLayerFeatures`. Let me know if the tests that I added are comprehensive enough. 
2. Updated the documentation for `pickImageryLayers` according to your specifications.
3. Removed an unnecessary `if` statement from `pickImageryLayers`.
4. Updated the helper function `pickImageryHelper`, removing it from the `ImageryLayerCollection` class.
5. Implemented your suggestion of using a callback function to minimize code duplication. Upon further consideration, I agree with this approach. I think it makes the code is a lot more intuitive and `ImageryLayerCollection.js` now matches the rest of our codebase more closely.

lilleyse commented on Specs/Scene/ImageryLayerCollectionSpec.js:
```diff
@@ -258,6 +258,260 @@ describe(
+       * Repeatedly calls update until the load queue is empty.  Returns a promise that resolves
+       * once the load queue is empty.
+       */
+      function updateUntilDone(globe) {
+        // update until the load queue is empty.
+        return pollToPromise(function () {
+          globe._surface._debug.enableDebugOutput = true;
+          scene.render();
+          return (
+            globe._surface.tileProvider.ready &&
+            globe._surface._tileLoadQueueHigh.length === 0 &&
+            globe._surface._tileLoadQueueMedium.length === 0 &&
+            globe._surface._tileLoadQueueLow.length === 0 &&
+            globe._surface._debug.tilesWaitingForChildren === 0
+          );
+        });
+      }
```
Is it possible to move this helper function outside of the `describe` block so that it can be used as a helper both here and  in `pickImageryLayerFeatures`? `scene` will need to be passed in as a parameter is all.

lilleyse commented on Specs/Scene/ImageryLayerCollectionSpec.js:
```diff
@@ -258,6 +258,260 @@ describe(
+        expect(imagery).toBeUndefined();
+      });
+
+      it("returns undefined when ImageryProvider does not implement pickFeatures", function () {
+        var provider = {
+          ready: true,
+          rectangle: Rectangle.MAX_VALUE,
+          tileWidth: 256,
+          tileHeight: 256,
+          maximumLevel: 0,
+          minimumLevel: 0,
+          tilingScheme: new GeographicTilingScheme(),
+          errorEvent: new Event(),
+          hasAlphaChannel: true,
+
+          requestImage: function (x, y, level) {
+            return ImageryProvider.loadImage(this, "Data/Images/Blue.png");
+          },
+        };
+
+        globe.imageryLayers.addImageryProvider(provider);
+
+        return updateUntilDone(globe).then(function () {
+          var ellipsoid = Ellipsoid.WGS84;
+          camera.lookAt(
+            new Cartesian3(ellipsoid.maximumRadius, 0.0, 0.0),
+            new Cartesian3(0.0, 0.0, 100.0)
+          );
+
+          var ray = new Ray(camera.position, camera.direction);
+          var imagery = scene.imageryLayers.pickImageryLayers(ray, scene);
+          expect(imagery).toBeUndefined();
+        });
+      });
```
This test should be removed because `pickFeatures` isn't really relevant to `pickImageryLayers`

CC https://github.com/CesiumGS/cesium/pull/9651#discussion_r667205720

lilleyse replied:
Same comment for the test below. It can be removed.

lilleyse replied:
For some reason these tests fail when run individually but pass when run with all `ImageryLayerCollection` tests. I would have expected them to fail in both cases. Could you check why this is happening? It's possible that previous tests are altering the scene in some way so that it always passes.

Passes: http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection

Fails: http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20undefined%20when%20ImageryProvider%20does%20not%20implement%20pickFeatures

lilleyse commented on Specs/Scene/ImageryLayerCollectionSpec.js:
```diff
@@ -258,6 +258,260 @@ describe(
+          errorEvent: new Event(),
+          hasAlphaChannel: true,
+
+          pickFeatures: function (x, y, level, longitude, latitude) {
+            var deferred = when.defer();
+            setTimeout(function () {
+              var featureInfo = new ImageryLayerFeatureInfo();
+              deferred.resolve([featureInfo]);
+            }, 1);
+            return deferred.promise;
+          },
```
Same comment: `pickFeatures` is not required for `pickImageryLayers` to work. It can be removed from these new tests.

lilleyse reviewed (commented):
Thanks @srothst1 - the main code looks good - I have some comments on the tests. Make sure to run code coverage (see https://github.com/CesiumGS/cesium/pull/9651#issuecomment-877432240) - there are some code paths that would be frequently hit that could be better tested. I marked some of those in red.

![Untitled](https://user-images.githubusercontent.com/915398/125544000-fece4fa0-2a5d-461d-ba2c-59b23dda9697.png)

srothst1 commented:
@lilleyse a few updates:

1. We now see that the [`ImageryLayerCollection`](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection) spec passes. In addition, the following individual specs in `pickImageryLayers` now pass:
- [returns undefined when pick ray does not intersect surface](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20undefined%20when%20pick%20ray%20does%20not%20intersect%20surface)
- [returns undefined when globe has no pickable layers](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20undefined%20when%20globe%20has%20no%20pickable%20layers)
- [returns imagery from one layer](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20imagery%20from%20one%20layer)
- [returns imagery from two layers](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20imagery%20from%20two%20layers)
- [returns undefined if there are zero imagery layers](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20returns%20undefined%20if%20there%20are%20zero%20imagery%20layers) (covers `imageryLayers.length === 0`)

2. I created a spec providing coverage for `imageryLayers.length === 0` in `pickImageryLayers`.

3. The following paths can be found in `pickImageryHelper`:

- line 356 -> undefined pickedTile
- line 370 -> `pickFeatures && !defined(provider.pickFeatures)`
- line 374 -> `!Rectangle.contains(imagery.rectangle, pickedLocation)`

 Should specs covering these paths go in a separate `describe`, or should they be implemented in "pickImageryLayers" or "pickImageryLayerFeatures"? Do you have any suggestions for how to cover these paths?

3. I moved the `updateUntilDone` helper function out of the `describe` block of "pickImageryLayers" and "pickImageryLayerFeatures". `updateUntilDone` is now a stand-alone function, making the code cleaner looking.

4. I removed tests from `pickImageryLayers` that referenced `ImageryProvider.pickFeatures` and `pickFeatures` references from "returns imagery from one layer" and "returns imagery from two layers" in "pickImageryLayers".

lilleyse commented:
> Should specs covering these paths go in a separate describe, or should they be implemented in "pickImageryLayers" or "pickImageryLayerFeatures"? Do you have any suggestions for how to cover these paths?

The specs can be put in group they're most relevant to. If there's overlap I think `pickImageryLayers` is fine because it's the more general capability. This is how I would do it:

`pickImageryLayers` describe block

- line 356 -> undefined pickedTile
- line 374 -> !Rectangle.contains(imagery.rectangle, pickedLocation)

`pickImageryLayerFeatures` describe block

- line 370 -> `pickFeatures && !defined(provider.pickFeatures)`

srothst1 commented:
@lilleyse I apologize for the delay - I just added three specs that should address the cases outlined above. 

- ["pickImageryHelper returns if undefined tile picked"](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20pickImageryHelper%20returns%20if%20undefined%20tile%20picked)
- ["continues if imagery rectangle does not contain picked location"](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20continues%20if%20imagery%20rectangle%20does%20not%20contain%20picked%20location)
- ["pickImageryHelper continues if provider.pickFeatures is undefined"](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayerFeatures%20pickImageryHelper%20continues%20if%20provider.pickFeatures%20is%20undefined)

["continues if imagery rectangle does not contain picked location"](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20continues%20if%20imagery%20rectangle%20does%20not%20contain%20picked%20location) covers the following code in `ImageryLayerCollection.js` when run individually (checked using DevTools)
```
    if (!Rectangle.contains(imagery.rectangle, pickedLocation)) {
      continue;
    }
```
However, coverage is not shown when running `npm run coverage`.

![image](https://user-images.githubusercontent.com/39531367/126554447-b389bb72-8abf-4f7b-9aa5-aa1dd2ee2aa0.png)

Any ideas of why we are seeing this behavior?

lilleyse reviewed (commented):
```
"pickImageryHelper returns if undefined tile picked"
"continues if imagery rectangle does not contain picked location"
```

These descriptions could be improved. They should be describing the behavior, not how the code is written. Maybe instead:

* pickImageryLayers returns undefined if no tiles are picked
* pickImageryLayers skips imagery layers that don't overlap the picked location

```
"pickImageryHelper continues if provider.pickFeatures is undefined"
 ```
 
 This ought to already be covered by `returns undefined when ImageryProvider does not implement pickFeatures`...

> ["continues if imagery rectangle does not contain picked location"](http://localhost:8080/Specs/SpecRunner.html?spec=Scene%2FImageryLayerCollection%20pickImageryLayers%20continues%20if%20imagery%20rectangle%20does%20not%20contain%20picked%20location) covers the following code in `ImageryLayerCollection.js` when run individually (checked using DevTools)
>...
> However, coverage is not shown when running `npm run coverage`.

This is most likely related to https://github.com/CesiumGS/cesium/pull/9651#discussion_r669199585. You'll need to get to the bottom of why. This could also explain why other tests aren't hitting the right code.

srothst1 commented on Specs/Scene/ImageryLayerCollectionSpec.js:
```diff
@@ -258,17 +258,52 @@ describe(
+        customGlobe = customScene.globe = new Globe();
+        customCamera = customScene.camera;
+
+        customScene.frameState.passes.render = true;
```
I originally added the set of custom variables because the spec `"pickImageryLayers returns undefined if no tiles are picked"` modifies `globe._surface._tilesToRender` and I do not want this change to impact the other specs.

lilleyse commented on Source/Scene/ImageryLayerCollection.js:
```diff
@@ -326,47 +326,17 @@ ImageryLayerCollection.prototype.lowerToBottom = function (layer) {
- * Asynchronously determines the imagery layer features that are intersected by a pick ray.  The intersected imagery
- * layer features are found by invoking {@link ImageryProvider#pickFeatures} for each imagery layer tile intersected
- * by the pick ray.  To compute a pick ray from a location on the screen, use {@link Camera.getPickRay}.
+ * Helper function for `pickImageryLayers` and `pickImageryLayerFeatures`.
```
It's fine to remove the documentation above `pickImageryHelper` since it's a local function.

Also some of the doc is misformatted (I think in `@return`) which is causing CI to fail, but no need to worry about since the doc will be removed.

srothst1 replied:
@lilleyse I just removed the documentation above `pickImageryHelper`. CI now passes.

lilleyse replied:
Thanks!

lilleyse reviewed (commented):
@srothst1 just one comment and then this PR can be merged.

lilleyse commented:
@srothst1 Sorry one last thing - could you add an entry to `CHANGES.md`?

srothst1 commented:
@lilleyse done!

lilleyse commented:
> - Added the ability to select imagery layers. [#9651](https://github.com/CesiumGS/cesium/pull/9651)

This could use a little more detail, maybe:

> - Added `ImageryLayerCollection.pickImageryLayers` which determines the imagery layers that are intersected by a pick ray. [#9651](https://github.com/CesiumGS/cesium/pull/9651)
